schema_based_on_diagram.sql

```diff
@@ -0,0 +1,13 @@
+CREATE DATABASE patient_clinic;
+CREATE TABLE patients(id INT GENERATED ALWAYS AS IDENTITY, name VARCHAR(100), date_of_birth DATE, PRIMARY KEY(id));
+CREATE TABLE medical_histories(id INT GENERATED ALWAYS AS IDENTITY, admitted_at TIMESTAMP, patient_id INT, status VARCHAR(100), PRIMARY KEY(id));
+ALTER TABLE medical_histories ADD CONSTRAINT patient FOREIGN KEY(patient_id) REFERENCES patients(id);
+CREATE TABLE invoices(id INT GENERATED ALWAYS AS IDENTITY, total_amount DECIMAL(10,2), generated_at TIMESTAMP, payed_at TIMESTAMP, medical_history_id INT, PRIMARY KEY(id));
+ALTER TABLE invoices ADD CONSTRAINT history FOREIGN KEY(medical_history_id) REFERENCES medical_histories(id);
+CREATE TABLE invoice_items(id INT GENERATED ALWAYS AS IDENTITY, unit_price DECIMAL(10,2), quantity INT, total_price DECIMAL(10,2), invoice_id INT, treatment_id INT, PRIMARY KEY(id));
+ALTER TABLE invoice_items ADD CONSTRAINT invoice FOREIGN KEY(invoice_id) REFERENCES invoices(id);
+CREATE TABLE treatments(id INT GENERATED ALWAYS AS IDENTITY, type VARCHAR(100), name VARCHAR(100), PRIMARY KEY(id));
+ALTER TABLE invoice_items ADD CONSTRAINT treatment FOREIGN KEY(treatment_id) REFERENCES treatments(id);
+CREATE TABLE histories_treatments(medical_history_id INT, treatment_id INT);
+ALTER TABLE histories_treatments ADD CONSTRAINT foreign_treatment FOREIGN KEY(treatment_id) REFERENCES treatments(id);
```

Comment:
- [ ] Kindly add explicit indexes.  These indexes are not constraints for maintaining data integrity but are used to speed up data retrieval when querying the database.

- [ ] Here are the specific indexes needed to be  created i:

- An index on the `patient_id` column of the `medical_histories` table.
An index on the `medical_history_id` column of the `invoices` table.
- Two separate indexes on the `invoice_id` and `treatment_id columns of the `invoice_items` table.
- Two separate indexes on the `medical_history_id` and `treatment_id columns` of the `histories_treatments` table.

### Sample code :
* `CREATE INDEX ON medical_histories (patient_id);`
* `CREATE INDEX ON invoices (medical_history_id);`
* `CREATE INDEX ON invoice_items (invoice_id);`
* `CREATE INDEX ON invoice_items (treatment_id);`
* `CREATE INDEX ON medical_histories_has_treatments (medical_history_id);`
* `CREATE INDEX ON medical_histories_has_treatments (treatment_id);`


- These explicit indexes help improve the performance of queries that involve filtering or sorting data based on the indexed columns. While foreign key constraints serve the purpose of maintaining data consistency, indexes are used to make database queries faster. It's common to have both foreign key constraints and indexes in a well-structured database to ensure both data integrity and query performance.

### Requirement:
![image](https://github.com/mahammad-mostafa/patient-clinic/assets/50990925/31f4aa33-9dd5-424b-aa9b-d7f1178ee3ee)